### DISCUSSION TITLE
modernize the singularity python module

### DISCUSSION THREAD

**Discussion by facastagnini** [singularity/singularity#330] [General]
Hi there!

I love this game and find myself coming back to it over the years, thanks for making this project open source, you rock 🥇 

I have worked on many python modules and found that this template is awesome [Hypermodern Python Cookiecutter](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/)

How open are we to do some refactoring and use those modern python module patterns on this project?
I am volunteering to do the work, I just want to test the waters to see if this is welcomed at all.

I would love to see github actions running CI tests on every PR, for example.

**Comment by nthykier** [comment]
> Hi there!
> 
> I love this game and find myself coming back to it over the years, thanks for making this project open source, you rock 1st_place_medal
> 
> I have worked on many python modules and found that this template is awesome [Hypermodern Python Cookiecutter](https://cookiecutter-hypermodern-python.readthedocs.io/en/2022.6.3.post1/)
> 
> How open are we to do some refactoring and use those modern python module patterns on this project? I am volunteering to do the work, I just want to test the waters to see if this is welcomed at all.
> 

This is a difficult question to answer:

1.  I know only most of the frameworks by name and not by how they work, it is hard for me to determine how open I am to a given of these tools. For me, it would be important that they provide value over their added "cost". As examples: 
   * I run `py.test` directly from the command line and using `nox` as an intermediate would only get in my way. This does not rule `nox` out being useful for other things (e.g., multi-python version CI stuff). In other words, some tools might need selling for why or how they are a benefit. It is probably a good idea to include the "cop-out" quick reference in the PR where relevant (e.g., there is always that chuck of code that needs to be hand formatted)
   * For typeguard, I am probably not interested in having typeguard on everything. On occasion, I do use duck typing and additional typeguard would come with a performance cost. Sometimes it makes sense, other times it does not.  For me, I feel you would have to show me a case in the singularity codebase where I would benefit from having typeguard - typeguard's own example of loading a json and checking it against a TypedDict is an interesting use case, that I could see value in (in other projects). But I am not aware of singularity having this. Though, refactoring our current data files from `.dat` to `.json` and using this pattern over the current things we have _might_ intrigue me. It is important for me that I know exactly which entry is wrong and also why it is wrong, so a "list entry 20 does not match the expected type" leaving you to count your way to the entry and then figure out what was wrong would be "meh" - flip side "Entry with name X is missing attribute Y", I am very much excited by things like that. But that is also a much bigger task than just "pip install typeguard" and sprinkling in some type checks or annotations, so I appreciate if that is more than you are willing to commit to.
2. There is very little activity on this project at the moment. I do not mind looking an occasional PR, but I recommend you do not over commit your efforts.  Obviously, if you are interested in developing this game further than just this, I am more than happy to review your changes with that in mind and eventually give you direct write access. Though for this, you should really have a vision of where you want to take the game - this is what I feel I am lacking right now and having two of us not knowing where to take the game does not really change anything. If this is your end goal, I am also more willing to accept tooling changes that I do not necessarily see the value in if they bring you value as I would expect you to lead development eventually.  Though, such a vision is probably a discussion in its own right.

With that said, migration from `setup.py` to `pyproject.toml` would be welcome as well (did not see it on the Cookiecutter list) assuming it is applicable to singularity. Not sure if it feature interacts with our window build tool though.

> I would love to see github actions running CI tests on every PR, for example.

We already have CI. Not sure if it stopped working for PRs though, that would be concerning and should definitely be fixed. A quick PR to fix if it is not working definitely has my go!


Moving on to more concrete tooling that I have already used in the past or I would probably be ok with from the cookier cutter list:

 * GitHub Action (we already use them here)
 * Flake8: Used a few things before in CI style rather than pre-commit. I probably have reservations about some of the checks from flake8 (historically, I agree with/accept 80-95% of verdicts with static analysis tools).
 * Py.test - already using as a test runner and for tests (at least in other projects; I do not remember of the tests we have here are py.tests or unit tests)
 * Coverage.py - already using it else where (and locally every now and then when working on tests).
 * Codecov - have used it in the past. To be honest, I prefer statically hosting the coverage.py report via gitlab pages on the gitlab projects I have to codecov. Not sure how to do that via GitHub though.
 * mypy - have used in one other project. It will have to be added to the CI if we are to use it here.
 * sphinx - already used for the (very minor amount of) developer docs we have. I do not mind having it for other docs as well. But to be honest, this will provide almost no value if the PR is just adding an "empty" shinx template with no docs in it. Such a PR would need to come with content in the documentation if you want to use it for user docs (etc.).  However, adding autodocs to the current developer docs might be fine
 * xdoctest - I never used it before, but I have used py.test --doctest-modules before, so I am fine with the concept of doctests (in fact, I prefer having doctests run during CI).  I prefer my doctest to also be included in my coverage report as sometimes I use doctests as the only tests for simple functions. Given that `py.test --doctest-modules` exist and work, I would probably need some convincing in that `xdoctest` does beyond that as it would be fighting against my finger memory (also, `--doctest-modules` can be enabled by the default in `pyproject.toml` file, which I probably should do in `singularity`).
 * dependabot - Used elsewhere. But usually only for projects with the .lock files or "strictly equal dependency support".  I have it enabled for singularity but only the "security alert" mode. Not sure it works though (never saw a dependabot alert on this repo, which can mean multiple things). Though, given our current set of dependencies, I would probably say this is a "low priority" thing for me.
 * black - Never used black before, but I have used code formatting tools in other languages before. Usually I am fine with that as long as the style is enforced (e.g., CI check) and there is a manual cop-out option for those 2 chucks of codes that formatter gets wrong. 



Tools that might be controversial/problematic:

 * The automatic upload via PyPI/TestPyPI.  Our PyPI account is managed via the original author, so that is not my decision to make. But also, it should probably also not be the first tool we adopt anyway.
 * Click. I am probably fine with trying this framework - but the CLI is "least" important aspect of singularity, so put it in the "last" bucket.
 * isort - I use pycharm and leave it to deal with my isort - sometimes I hit "origanize imports" and other times not. If pycharm automatically picks up ispell and runs it for me, ok. If it will be another tool that I need to run - not so much.  Note that I willing to run a general code formatter (e.g., black) and if that covers the isort part, then that works for me. Though then isort is redundant for me - I would use black, not isort (so unless black depends on isort for this purpose, I would do away with the isort dependency).
 * Any "mandatory" local development or release tool that is not in Debian. Virtualenv has not really been my style so far and that worked out for singularity up to now (probably also why I have not been using nox nor tox). It would take me a lot of convincing to change this (not saying it is impossible, but I cannot recommend opening the ball here). I am less concerned here about things that only needs to run on GitHub CI - it can `pip install` away. Though, I suspect this in practice will be a non-issue (at least mypy, flake, pytest, sphinx, and black are in Debian, so we are just discussing version alignment for those)


With this in mind, feel free to send PRs. I would recommend you start with low hanging non-controversial fruits - I would hate for you to work on a multi-hour/day change only for me to say "That one does not work for me" that would be a bummer.  There has not been a lot of activity in singularity, but I will prioritize having a look at your tooling PRs if you decide to do them. :)

**Reply by nthykier** [reply]
Note I would be interested in a refactoring from singularity's own homegrown UI into a proper UI framework like https://github.com/MyreMylar/pygame_gui (even if that is not packaged in Debian yet). It has nothing to do with the cookiecutter tool list, but I thought I would mention it if you are interested in general refactoring as well.

**Comment by facastagnini** [comment]
@nthykier thanks a lot for the reply, I am enjoying some time off, I will be able to put some time in a few weeks.

As a quick reply right now, I am more than happy giving you PRs, no ambition to get write access :P

I agree with the list of tools you mentioned, and since `pyproject.toml` allows to centralize the configuration for most if not all of them, I would like to start migrating from `setup.py` to `pyproject.toml`, then we can add tools one at the time.
You mentioned the windows build, we will be in touch over the PR until the change is smooth.

I am all for small incremental changes, nothing dramatic like commingled updates.

**Reply by nthykier** [reply]
Sounds good. :)

I think the Windows build tool is `py2exe` based on https://github.com/singularity/singularity/blob/master/release/README.txt. I never used it myself and I would not know how to verify whether it works.

But indeed, starting with pyproject.toml is fine. :) Looking forward to hearing from you. :)
